File: src/main/services/ModActivator.ts
import fs from 'fs';
import path from 'path';
import type { ModManifestDto, ModMetadataDto } from '../../shared/types';
import { DISABLED_EXTENSION, bepinexPatchersDir } from '../constants';
import { loadMetadata, saveMetadata } from './ManifestLoader';
const pluginExt = '.dll';
const DISABLED_SUFFIX = `.${DISABLED_EXTENSION}`;
function dllStem(relative: string): string {
    let s = relative;
    let lc = s.toLowerCase();
    if (lc.endsWith('.disabled'))
        s = s.slice(0, -9);
    else if (lc.endsWith('.disable'))
        s = s.slice(0, -8);
    lc = s.toLowerCase();
    const backup = /^(.*)\.dll\.\d+$/i.exec(s);
    if (backup)
        return backup[1];
    if (lc.endsWith('.dll'))
        return s.slice(0, -4);
    return s;
}
function disabledCandidates(base: string): string[] {
    const stem = path.basename(base);
    return [
        `${stem}${DISABLED_SUFFIX}`,
        `${stem}.disable`,
        `${stem}.dll${DISABLED_SUFFIX}`,
        `${stem}.dll.disable`,
        `${stem}.dll.1`
    ];
}
function detectStemFromName(name: string): string | null {
    if (!name || name[0] === '.')
        return null;
    const lc = name.toLowerCase();
    if (lc.endsWith(DISABLED_SUFFIX)) {
        const raw = name.slice(0, -DISABLED_SUFFIX.length);
        return raw.toLowerCase().endsWith('.dll') ? raw.slice(0, -4) : raw;
    }
    if (lc.endsWith('.disable')) {
        const raw = name.slice(0, -8);
        return raw.toLowerCase().endsWith('.dll') ? raw.slice(0, -4) : raw;
    }
    if (lc === '')
        return null;
    const dot = name.lastIndexOf('.');
    return dot > 0 ? name.slice(0, dot) : null;
}
function findDisabledFile(dir: string, base: string): string | null {
    const stem = path.basename(base).toLowerCase();
    let found: string | null = null;
    for (const f of fs.readdirSync(dir)) {
        if (f.toLowerCase() === `${stem}.dll`)
            continue;
        const detected = detectStemFromName(f);
        if (detected && detected.toLowerCase() === stem) {
            if (found)
                return null;
            found = f;
        }
    }
    return found;
}
function renameDll(dir: string, base: string, activate: boolean): boolean {
    const active = path.join(dir, `${path.basename(base)}.dll`);
    if (activate) {
        if (fs.existsSync(active))
            return true;
        for (const cand of disabledCandidates(base)) {
            const src = path.join(dir, cand);
            if (fs.existsSync(src) && fs.statSync(src).isFile()) {
                fs.renameSync(src, active);
                return true;
            }
        }
        const stemName = path.basename(base).toLowerCase();
        for (const f of fs.readdirSync(dir)) {
            if (/\.dll\.\d+$/i.test(f) && f.toLowerCase().startsWith(`${stemName}.dll.`)) {
                fs.renameSync(path.join(dir, f), active);
                return true;
            }
        }
        const loose = findDisabledFile(dir, base);
        if (loose && loose.toLowerCase() !== `${path.basename(base)}.dll`) {
            fs.renameSync(path.join(dir, loose), active);
            return true;
        }
        return false;
    }
    if (fs.existsSync(active)) {
        fs.renameSync(active, path.join(dir, `${path.basename(base)}${DISABLED_SUFFIX}`));
        return true;
    }
    return true;
}
export function toggleActivation(gameRoot: string, modRoot: string, manifest: ModManifestDto, activate: boolean): {
    activated: boolean;
} {
    for (const p of manifest.plugins) {
        const pluginPath = path.resolve(modRoot, p.startsWith('..') ? path.join('..', p) : p);
        const dir = path.dirname(pluginPath);
        const base = path.basename(pluginPath, pluginExt);
        if (fs.existsSync(dir))
            renameDll(dir, base, activate);
    }
    const patchersDir = bepinexPatchersDir(gameRoot);
    for (const p of manifest.patchers) {
        const base = path.basename(p, pluginExt);
        if (fs.existsSync(patchersDir))
            renameDll(patchersDir, base, activate);
    }
    const meta: ModMetadataDto = { ...loadMetadata(modRoot, manifest), activated: activate };
    saveMetadata(modRoot, manifest, meta);
    return { activated: activate };
}
export function toggleLegacyPlugin(installDir: string, pluginFiles: string[], activate: boolean): {
    activated: boolean;
    pluginFiles: string[];
} {
    const nextPluginFiles: string[] = [];
    for (const pf of pluginFiles) {
        const dir = path.dirname(path.resolve(installDir, pf));
        renameDll(dir, dllStem(pf), activate);
        const logical = `${path.basename(dllStem(pf))}${pluginExt}`;
        const prefix = path.dirname(pf);
        nextPluginFiles.push(prefix === '.' ? logical : path.join(prefix, logical));
    }
    return { activated: activate, pluginFiles: nextPluginFiles };
}

File: RBBPMM.Wpf/RBBPMM.Core/ModActivator.cs
using System.Text.RegularExpressions;

namespace RBBPMM.Core;

/// <summary>Enable/disable mods by renaming their DLLs to/from the <c>.disabled</c> sentinel.</summary>
public static class ModActivator
{
    private const string PluginExt = ".dll";
    private const string DisabledSuffix = ".disabled";

    /// <summary>Strip activation suffixes and extensions to get a DLL's logical stem.</summary>
    public static string DllStem(string relative)
    {
        var s = relative;
        var lc = s.ToLowerInvariant();
        if (lc.EndsWith(".disabled")) s = s[..^9];
        else if (lc.EndsWith(".disable")) s = s[..^8];
        lc = s.ToLowerInvariant();
        var backup = Regex.Match(s, @"^(.*)\.dll\.\d+$", RegexOptions.IgnoreCase);
        if (backup.Success) return backup.Groups[1].Value;
        if (lc.EndsWith(".dll")) return s[..^4];
        return s;
    }

    private static List<string> DisabledCandidates(string baseName)
    {
        var stem = Path.GetFileName(baseName);
        return
        [
            $"{stem}{DisabledSuffix}",
            $"{stem}.disable",
            $"{stem}.dll{DisabledSuffix}",
            $"{stem}.dll.disable",
            $"{stem}.dll.1"
        ];
    }

    private static string? DetectStemFromName(string name)
    {
        if (string.IsNullOrEmpty(name) || name[0] == '.') return null;
        var lc = name.ToLowerInvariant();
        if (lc.EndsWith(DisabledSuffix))
        {
            var raw = name[..^DisabledSuffix.Length];
            return raw.ToLowerInvariant().EndsWith(".dll") ? raw[..^4] : raw;
        }
        if (lc.EndsWith(".disable"))
        {
            var raw = name[..^8];
            return raw.ToLowerInvariant().EndsWith(".dll") ? raw[..^4] : raw;
        }
        var dot = name.LastIndexOf('.');
        return dot > 0 ? name[..dot] : null;
    }

    private static string? FindDisabledFile(string dir, string baseName)
    {
        var stem = Path.GetFileName(baseName).ToLowerInvariant();
        string? found = null;
        foreach (var f in Directory.GetFiles(dir))
        {
            var fn = Path.GetFileName(f);
            if (fn.ToLowerInvariant() == $"{stem}.dll") continue;
            var detected = DetectStemFromName(fn);
            if (detected != null && detected.ToLowerInvariant() == stem)
            {
                if (found != null) return null;
                found = fn;
            }
        }
        return found;
    }

    private static bool RenameDll(string dir, string baseName, bool activate)
    {
        var active = Path.Combine(dir, $"{Path.GetFileName(baseName)}.dll");
        if (activate)
        {
            if (File.Exists(active)) return true;
            foreach (var cand in DisabledCandidates(baseName))
            {
                var src = Path.Combine(dir, cand);
                if (File.Exists(src) && !Directory.Exists(src))
                {
                    File.Move(src, active);
                    return true;
                }
            }
            var stemName = Path.GetFileName(baseName).ToLowerInvariant();
            foreach (var f in Directory.GetFiles(dir))
            {
                var fn = Path.GetFileName(f);
                if (Regex.IsMatch(fn, @"\.dll\.\d+$", RegexOptions.IgnoreCase) &&
                    fn.ToLowerInvariant().StartsWith($"{stemName}.dll."))
                {
                    File.Move(f, active);
                    return true;
                }
            }
            var loose = FindDisabledFile(dir, baseName);
            if (loose != null && !loose.Equals($"{Path.GetFileName(baseName)}.dll", StringComparison.OrdinalIgnoreCase))
            {
                File.Move(Path.Combine(dir, loose), active);
                return true;
            }
            return false;
        }
        if (File.Exists(active))
        {
            File.Move(active, Path.Combine(dir, $"{Path.GetFileName(baseName)}{DisabledSuffix}"));
            return true;
        }
        return true;
    }

    public static bool ToggleActivation(string gameRoot, string modRoot, ModManifest manifest, bool activate)
    {
        foreach (var p in manifest.Plugins)
        {
            var pluginPath = Path.GetFullPath(Path.Combine(modRoot, p.StartsWith("..") ? Path.Combine("..", p) : p));
            var dir = Path.GetDirectoryName(pluginPath)!;
            var baseName = Path.GetFileNameWithoutExtension(pluginPath);
            if (Directory.Exists(dir))
                RenameDll(dir, baseName, activate);
        }
        var patchersDir = Constants.BepinexPatchersDir(gameRoot);
        foreach (var p in manifest.Patchers)
        {
            var baseName = Path.GetFileNameWithoutExtension(p);
            if (Directory.Exists(patchersDir))
                RenameDll(patchersDir, baseName, activate);
        }
        var meta = ManifestLoader.LoadMetadata(modRoot);
        meta.Activated = activate;
        ManifestLoader.SaveMetadata(modRoot, meta);
        return activate;
    }

    public static (bool Activated, List<string> PluginFiles) ToggleLegacyPlugin(
        string installDir, List<string> pluginFiles, bool activate)
    {
        var next = new List<string>();
        foreach (var pf in pluginFiles)
        {
            var dir = Path.GetDirectoryName(Path.GetFullPath(Path.Combine(installDir, pf)))!;
            RenameDll(dir, DllStem(pf), activate);
            var logical = $"{Path.GetFileName(DllStem(pf))}{PluginExt}";
            var prefix = Path.GetDirectoryName(pf) ?? "";
            next.Add(prefix == "." ? logical : Path.Combine(prefix, logical));
        }
        return (activate, next);
    }
}
